File: DSA/GreedyAlgo/Code/CoinChange.java
// Coin Change Problem in Java

import java.util.*;

public class CoinChange {
    public static void coinChange(int[] coins, int amount) {
        Arrays.sort(coins);
        
        System.out.println("Coin Change for amount " + amount + ":");
        int coinsUsed = 0;
        
        for (int i = coins.length - 1; i >= 0 && amount > 0; i--) {
            while (amount >= coins[i]) {
                System.out.println("Take coin of value " + coins[i]);
                amount -= coins[i];
                coinsUsed++;
            }
        }
        
        System.out.println("\nTotal Coins Used: " + coinsUsed);
        if (amount > 0) {
            System.out.println("Warning: Cannot make exact change (Remaining: " + amount + ")");
        } else {
            System.out.println("Amount Made: Complete");
        }
    }
    
    public static void main(String[] args) {
        System.out.println("=== Coin Change Problem (Java) ===\n");
        
        int[] coins = {1, 5, 10, 25};
        int amount = 41;
        
        System.out.print("Available Coins: ");
        for (int coin : coins) {
            System.out.print(coin + " ");
        }
        System.out.println("\nAmount to Make: " + amount + "\n");
        
        coinChange(coins, amount);
        
        System.out.println("\n=== Greedy Strategy ===");
        System.out.println("1. Sort coins in descending order");
        System.out.println("2. For each coin (largest to smallest):");
        System.out.println("   - Use as many as possible");
        System.out.println("   - Subtract from amount");
        System.out.println("3. Continue until amount becomes 0\n");
        
        System.out.println("=== Complexity ===");
        System.out.println("Time Complexity:  O(n log n) for sorting + O(amount)");
        System.out.println("Space Complexity: O(1)");
        System.out.println("Note: Greedy works for standard coin systems");
    }
}

File: DSA/GreedyAlgo/Code/CoinChange.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void coinChange(vector<int>& coins, int amount) {
    sort(coins.rbegin(), coins.rend());
    
    cout << "Coin Change for amount " << amount << ":\n";
    int coinsUsed = 0;
    
    for (int coin : coins) {
        while (amount >= coin) {
            cout << "Take coin of value " << coin << "\n";
            amount -= coin;
            coinsUsed++;
        }
    }
    
    cout << "\nTotal Coins Used: " << coinsUsed << "\n";
    if (amount > 0) {
        cout << "Warning: Cannot make exact change (Remaining: " << amount << ")\n";
    } else {
        cout << "Amount Made: Complete\n";
    }
}

int main() {
    cout << "=== Coin Change Problem (C++) ===\n\n";
    
    vector<int> coins = {1, 5, 10, 25};
    int amount = 41;
    
    cout << "Available Coins: ";
    for (int coin : coins) {
        cout << coin << " ";
    }
    cout << "\nAmount to Make: " << amount << "\n\n";
    
    coinChange(coins, amount);
    
    cout << "\n=== Greedy Strategy ===\n";
    cout << "1. Sort coins in descending order\n";
    cout << "2. For each coin (largest to smallest):\n";
    cout << "   - Use as many as possible\n";
    cout << "   - Subtract from amount\n";
    cout << "3. Continue until amount becomes 0\n\n";
    
    cout << "=== Complexity ===\n";
    cout << "Time Complexity:  O(n log n) for sorting + O(amount)\n";
    cout << "Space Complexity: O(1)\n";
    cout << "Note: Greedy works for standard coin systems\n";
    
    return 0;
}
